docs/docs/icicle/core.md

```diff
@@ -0,0 +1,172 @@
+
+```sh
+cd icicle
+mkdir -p build
+cmake -DCURVE=<CURVE> -S . -B build
+cmake --build build -j
+```
+
+Where `<CURVE>` can be one of `bn254`/`bls12_377`/`bls12_381`/`bw6_761`/`grumpkin`.
+
+This command will output both `libingo_curve_<CURVE>.a` and `libingo_field_<CURVE>.a` into `build/lib`.
+
+### Compilation options
+
+There exist multiple options that allow you to customize your build or enable additional functionality.
+
+#### EXT_FIELD
+
+Used only in a [field mode](#compiling-for-a-field) to add Extension field into a build. Adds NTT for the extension field.
+
+Default: `OFF`
+
+Usage: `-DEXT_FIELD=ON`
+
+#### G2
+
+Used only in a [curve mode](#compiling-for-a-curve) to add G2 definitions into a build. Also adds G2 MSM.
+
+Default: `OFF`
+
+Usage: `-DG2=ON`
+
+#### ECNTT
+
+Used only in a [curve mode](#compiling-for-a-curve) to add ECNTT function into a build.
+
+Default: `OFF`
+
+Usage: `-DECNTT=ON`
+
+#### MSM
+
+Used only in a [curve mode](#compiling-for-a-curve) to add MSM function into a build. As MSM takes a lot of time to build, you can disable it with this option to reduce compilation time.
+
+Default: `ON`
+
+Usage: `-DMSM=OFF`
+
+#### BUILD_HASH
+
+Can be used in any mode to build a hash library. Currently it only includes Keccak hash function, but more are coming.
+
+Default: `OFF`
+
+Usage: `-DBUILD_HASH=ON`
+
+#### BUILD_TESTS
+
+Can be used in any mode to include tests runner binary.
+
+Default: `OFF`
+
+USAGE: `-DBUILD_TESTS=ON`
+
+### BUILD_BENCHMARKS
+
+Can be used in any mode to include benchmarks runner binary.
+
+Default: `OFF`
+
+USAGE: `-DBUILD_BENCHMARKS=ON`
+
+## Linking
+
+To link ICICLE with your project you first need to compile ICICLE with options of your choice. After that you can use CMake `target_link_libraries` to link with the generated static libraries and `target_include_directories` to include ICICLE headers (located in `icicle/include`).
+
+Refer to our [c++ examples](https://github.com/ingonyama-zk/icicle/tree/main/examples/c%2B%2B) for more info. Take a look at this [CMakeLists.txt](https://github.com/ingonyama-zk/icicle/blob/main/examples/c%2B%2B/msm/CMakeLists.txt#L22)
+
+
+## Writing new bindings for ICICLE
+
+Since ICICLE Core is written in CUDA / C++ its really simple to generate static libraries. These static libraries can be installed on any system and called by higher level languages such as Golang.
+
+Static libraries can be loaded into memory once and used by multiple programs, reducing memory usage and potentially improving performance. They also allow you to separate functionality into distinct modules so your static library may need to compile only specific features that you want to use.
+
+Lets review the Golang bindings since its a pretty verbose example (compared to rust which hides it pretty well) of using static libraries. Golang has a library named `CGO` which can be used to link static libraries. Here's a basic example on how you can use cgo to link these libraries:
```

Comment:
```suggestion
Let's review the [Golang bindings](golang-bindings.md) since its a pretty verbose example (compared to rust which hides it pretty well) of using static libraries. Golang has a library named `CGO` which can be used to link static libraries. Here's a basic example on how you can use cgo to link these libraries:
```